cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/WorkflowDataset.java

```diff
@@ -204,6 +207,78 @@ public int compare(WorkflowDataset.WorkflowRunRecord o1, WorkflowDataset.Workflo
     return programToRunRecord.values();
   }
 
+  @Nullable
+  public WorkflowRunRecord getRecord(Id.Workflow id, String runId) {
+    RunId pid = RunIds.fromString(runId);
+    long startTime = RunIds.getTime(pid, TimeUnit.SECONDS);
+    MDSKey mdsKey = new MDSKey.Builder().add(id.getNamespaceId())
+      .add(id.getApplicationId()).add(id.getId()).add(startTime).build();
+    byte[] startRowKey = mdsKey.getKey();
+    Scan scan = new Scan(startRowKey, null);
+
+    WorkflowRunRecord workflowRunRecord;
+    Scanner scanner = table.scan(scan);
+    Row indexRow = scanner.next();
+    if (indexRow == null) {
+      return null;
+    }
+    Map<byte[], byte[]> columns = indexRow.getColumns();
+    String workflowRunId = Bytes.toString(columns.get(RUNID));
+    long timeTaken = Bytes.toLong(columns.get(TIME_TAKEN));
+
+    List<ProgramRun> actionRunsList = GSON.fromJson(Bytes.toString(columns.get(NODES)), PROGRAM_RUNS_TYPE);
+    workflowRunRecord = new WorkflowRunRecord(workflowRunId, timeTaken, actionRunsList);
+
+    if (!runId.equals(workflowRunRecord.getWorkflowRunId())) {
+      return null;
+    } else {
+      return workflowRunRecord;
+    }
+  }
+
+  Map<String, WorkflowRunRecord> getDetailsOfRange(Id.Workflow workflow, String runId, int count, long timeInterval) {
+    Map<String, WorkflowRunRecord> mainRunRecords = getNeighbors(workflow, runId, count, timeInterval);
+    WorkflowRunRecord workflowRunRecord = getRecord(workflow, runId);
+    if (workflowRunRecord != null) {
+      mainRunRecords.put(workflowRunRecord.getWorkflowRunId(), workflowRunRecord);
+    }
+    return mainRunRecords;
+  }
+
+  private Map<String, WorkflowRunRecord> getNeighbors(Id.Workflow id, String runId, int count, long timeInterval) {
+    RunId pid = RunIds.fromString(runId);
+    long startTime = RunIds.getTime(pid, TimeUnit.SECONDS);
+    Map<String, WorkflowRunRecord> workflowRunRecords = new HashMap<>();
+    for (int i = (-1 * count); i <= count; i++) {
```

Comment:
this loop can we made better. use a while loop and instead of using ``prevStartTime = startTime + (i * timeInterval)``, update prevStartTime, based on the row what we find in scan, to avoid redoing some scan. you can instead do
``prevStartTime = time(L#268) + timeInterval``